#11 : Implement Light and Dark Theme System

leo7418 opened the pull request:
#11

leo7418 committed:
- Add light theme
- New class ThemeManager
- Refactor components for using theme instead of hardcoded colors
- Use latest TGUI version

f-michaut commented on include/Components/Foldout.hpp:
```diff
@@ -3,8 +3,8 @@
 **
-** Started on  Sat Jan 15 21:08:44 2022 Léo Lhuile
-** Last update Wed Sep 14 22:26:27 2022 Léo Lhuile
+** Started on  Sat Jan 15 21:08:44 2024 Léo Lhuile
+** Last update Wed Sep 14 22:26:27 2024 Léo Lhuile
```
The `Started on` year shouldn't be updated :see_no_evil:

leo7418 replied:
Actually, the 2024 date is correct here ; when I created this file, it was in 2024, not 2022 😄

f-michaut commented on include/Application.hpp:
Since there isn't any real changes in this file, could you revert it ?

f-michaut commented on include/Components/ListMenu.hpp:
```diff
@@ -40,7 +40,7 @@ namespace FileShare::GUI::Components {
             tgui::SignalString onMenuActive = { "MenuActived" };
             tgui::SignalString onMenuInactive = { "MenuInactived" };
-            tgui::SignalTyped<const std::vector<const tgui::String>> onSelectionChanged = { "SelectionChanged" };
+            tgui::SignalTyped<const std::vector<tgui::String>&> onSelectionChanged = { "SelectionChanged" };
```
```suggestion
            tgui::SignalTyped<const std::vector<tgui::String> &> onSelectionChanged = { "SelectionChanged" };
```

f-michaut commented on include/ThemeManager.hpp:
```diff
@@ -0,0 +1,35 @@
+            ThemeManager(const ThemeManager&) = delete;
+            ThemeManager &operator=(const ThemeManager&) = delete;
+
+            std::string currentTheme = "dark";
```
Shouldn't it default to `system`, and then fallback to `black` if can't determine system's theme ?

f-michaut commented on source/ThemeManager.cpp:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    void ThemeManager::setTheme(const std::string &theme)
+    {
```
project style is to have opening braces on same line :stuck_out_tongue: 
```suggestion
    void ThemeManager::setTheme(const std::string &theme) {
```

f-michaut commented on source/ThemeManager.cpp:
```diff
@@ -0,0 +1,31 @@
+        this->currentTheme = theme;
+        tgui::Theme::setDefault("assets/themes/" + theme + ".txt");
+    }
+}
```
And to have final newlines. Both are configured in the `.editorconfig` file - do you not have an Editorconfig plugin enabled in your editor ? Maybe check if its configured properly

f-michaut commented on include/ThemeManager.hpp:
```diff
@@ -0,0 +1,35 @@
+namespace FileShare::GUI {
+    class ThemeManager {
+        public:
+            static ThemeManager &getInstance();
```
This works, but how about also providing conveniance `static` functions for all the instance methods so instead of `ThemeManager::getInstance().getCurrentTheme()` we can also just do `ThemeManager::getCurrentTheme()` ?

f-michaut commented on include/ThemeManager.hpp:
```diff
@@ -0,0 +1,35 @@
+            void setTheme(const std::string &theme);
+            std::string getCurrentTheme() const { return this->currentTheme; }
+
+            std::string getIconThemeSuffix() const { return this->currentTheme == "light" ? "_black" : "_white"; }
```
Propositon, instead of having to do `"file_path" + suffix + ".extension" everywhere we use this`, could we have something like : 
```suggestion
            std::string getThemedAssetPath(std::string filename, std::string extension = ".svg") const { return "assets/images" + filename + (this->currentTheme == "light" ? "_black" : "_white") + extension; }
```

f-michaut replied:
also, I believe we should move the images in their own theme folder (eg: `assets/images/light/` and `assets/images/dark/`) and we can rename them to remove the `_white` / `_black` part.

Making sure we use the name of the theme for the folder, makes it so that we dont need that `light -> white / black -> dark` ternary.

f-michaut commented on source/Settings/View.cpp:
```diff
@@ -21,7 +22,7 @@ namespace FileShare::GUI::Settings {
         this->setAutoLayout(tgui::AutoLayout::Top);
         this->setAutoSeparatorsBeforeTitles();
         this->setAutoHeight(true);
-        this->getRenderer()->setPadding({6, 6});
+        this->getRenderer()->setPadding({ 6, 6 });
```
Codestyle-wise, I prefer without spaces :stuck_out_tongue:

f-michaut commented on source/Settings/View.cpp:
```diff
@@ -56,14 +59,14 @@ namespace FileShare::GUI::Settings {
         logoutButton->setText("Logout");
         logoutButton->setType(Components::Button::Type::Danger);
-        logoutButton->onPress([=]() {
+        logoutButton->onPress([this]() {
             this->getSignal("logout").emit(this); // TODO: Implement signal. For now, it's not a bug; it's a feature
-        });
+            });
```
Was good before - should be aligned with the begining of the line that opened them.

f-michaut commented on source/Components/ListMenu.cpp:
```diff
@@ -211,7 +211,7 @@ namespace FileShare::GUI::Components {
         size_t activeSubItemsSize = this->activeSubItems.size();
 
-        std::vector<const tgui::String> selection;
+        std::vector<tgui::String> selection;
```
Curious - what was the reason for removing the const ? Was it causing issues ?
I dont think we would need to modify the strings of the selection, so would makes sense that they are const imo ?

I would even wonder if we could have `const &` cause then it would avoid a copy-construction
(note, it would need to be `std::vector<std::reference_wrapper<const tgui::String>>>` since you cant make a vector of references directly)

f-michaut commented on source/Application.cpp:
```diff
@@ -23,18 +24,26 @@ namespace FileShare::GUI {
 
         tgui::ScrollablePanel::Ptr content = tgui::ScrollablePanel::create();
         content->setAutoLayout(tgui::AutoLayout::Fill);
+        if (themeManager.getCurrentTheme() == "dark") {
+            content->getRenderer()->setBackgroundColor(tgui::Color(35, 35, 35));
+        }
```
Hum, why do we have to set this here on `init` ?

Are theses things gonna update properly when we switch themes ?
If so, they should be setup properly without us having to do anything ?

f-michaut commented on source/ThemeManager.cpp:
```diff
@@ -0,0 +1,31 @@
+#include "ThemeManager.hpp"
+
+namespace FileShare::GUI {
+    ThemeManager::ThemeManager()
+    {
+        this->setTheme(this->currentTheme);
+    }
```
This does not seem persistent, but I forgot - do we currently persist settings or not yet ?

f-michaut reviewed (commented):
So far, changes makes sense to me. 

I'm gona try to pull theses changes tomorow and play with it and see if I can find any issue.

f-michaut commented on assets/themes/light.txt:
The input widgets are still Dark when in light mode :

<img width="790" height="322" alt="Image" src="https://github.com/user-attachments/assets/dda36a04-a0e6-42b3-bf0c-a60687f38500" />

leo7418 committed:
- Updated ThemeManager to support system theme detection and initialization.
- Refactored components to utilize themed assets instead of hardcoded colors.
- Add some rules in .editorconfig

f-michaut commented on .editorconfig:
```diff
@@ -5,7 +5,34 @@ end_of_line = lf
 indent_size = 4
 indent_style = space
+
+# Accolades et nouvelles lignes
```
English comments pls ^^

f-michaut commented on include/ThemeManager.hpp:
```diff
@@ -0,0 +1,48 @@
+#include <cstdlib>
+#include <fstream>
+#include <algorithm>
+#endif
```
I believe that indenting preprocessor #if makes them much more readable.
Also, instead of relying on the raw OS detection macros, we have `<CppSockets/OSDetection.hpp>` which defines `OS_LINUX`, `OS_WINDOWS`, and `OS_APPLE` which are easier to remember

```suggestion
<CppSockets/OSDetection.hpp>

#ifdef OS_WINDOWS
  #include <windows.h>
  #include <winreg.h>
#elif OS_APPLE
  #include <CoreFoundation/CoreFoundation.h>
#elif OS_LINUX
  #include <cstdlib>
  #include <fstream>
  #include <algorithm>
#endif
```

f-michaut commented on source/Application.cpp:
```diff
@@ -4,37 +4,38 @@
-    {
+        : window(sf::VideoMode(1200, 800), "FileShare")
+        , gui(window) {
     }
```
I know we where talking about style earlier, and I said all `{` are on the same line, but kinda forgot about the constructors :sweat_smile: 

That's an exception : 
```suggestion
    Application::Application() : 
      window(sf::VideoMode(1200, 800), "FileShare"), 
      gui(window)
    {}
```
If there is nothing in the `{}` they go alone on their line. If there is something, then : 
```
    Application::Application() : 
      window(sf::VideoMode(1200, 800), "FileShare"), 
      gui(window)
    {
      something();
    }
```

f-michaut commented on source/Components/List.cpp:
```diff
@@ -43,16 +39,15 @@ namespace FileShare::GUI::Components {
-                this->updateItemsPosition(widgetIndex);
-                this->updateHeight();
-            })
+            this->updateItemsPosition(widgetIndex);
+            this->updateHeight();
+        })
         );
```
Was good before :sweat_smile: 
```suggestion
                this->updateItemsPosition(widgetIndex);
                this->updateHeight();
            })
        );
```

f-michaut commented on source/ThemeManager.cpp:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    ThemeManager &ThemeManager::getInstance() {
+        static ThemeManager instance;
+        return instance;
+    }
```
```suggestion
        static ThemeManager instance;

        return instance;
    }
```

f-michaut commented on source/ThemeManager.cpp:
```diff
@@ -0,0 +1,145 @@
+
+    // TODO: remove after we store the value inside SQLite
+    void ThemeManager::initializeTheme() {
+        std::string actualTheme = this->currentTheme;
+
+        if (this->currentTheme == "system") {
+            actualTheme = detectSystemTheme();
+        }
+
+        this->currentTheme = actualTheme;
+        tgui::Theme::setDefault("assets/themes/" + actualTheme + ".txt");
```
Redundant logic with `setTheme()`
```suggestion
        ThemeManager::setTheme(this->currentTheme);
```

f-michaut commented on source/ThemeManager.cpp:
```diff
@@ -0,0 +1,145 @@
+                std::string scheme(buffer);
+                pclose(pipe);
+                std::transform(scheme.begin(), scheme.end(), scheme.begin(), ::tolower);
+                if (scheme.find("dark") != std::string::npos || scheme.find("breeze dark") != std::string::npos) {
```
Finding `"dark"` would have already triggered if its `"breeze dark"`
```suggestion
                if (scheme.find("dark") != std::string::npos) {
```

leo7418 committed:
- Fix some style guide issues
